n_serial.c

```diff
@@ -220,3 +162,106 @@ bool serialNoteReset()
     // Done
     return notecardReady;
 }
+
+/**************************************************************************/
+/*!
+  @brief  Receive bytes over Serial to the Notecard.
+  @param   buffer
+            A buffer to receive bytes into.
+  @param   size (in/out)
+            - (in) The size of the buffer in bytes.
+            - (out) The length of the received data in bytes.
+  @param   delay
+            Respect delay standard transmission delays.
+  @param   timeoutMs
+            The maximum amount of time, in milliseconds, to wait for serial data
+            to arrive. Passing zero (0) disables the timeout.
+  @param   overflow (out)
+            The overflow amount in bytes.
+  @returns  A c-string with an error, or `NULL` if no error ocurred.
+*/
+/**************************************************************************/
+const char *serialRawReceive(uint8_t *buffer, size_t *size, bool delay, size_t timeoutMs, size_t *overflow)
+{
+    // Check for special case(s)
+    if (!timeoutMs) {
+        timeoutMs = (size_t)-1;
+    }
+
+    size_t received = 0;
+    *overflow = (received >= *size);
+    const size_t startMs = _GetMs();
+    for (bool eop = false ; !(*overflow) && !eop ;) {
+        while (!_SerialAvailable()) {
+            if (_GetMs() - startMs >= timeoutMs) {
+                *size = received;
+                buffer[received] = '\0';
+#ifdef ERRDBG
+                _Debug("received only partial reply after timeout:\n");
+                _Debug((char *)buffer);
+                _Debug("^^ partial buffer contents ^^\n");
+#endif
+                return ERRSTR("transaction incomplete {io}",c_iotimeout);
+            }
+            if (delay) {
+                _DelayMs(1);
+            }
+        }
+        char ch = _SerialReceive();
+
+        // Place into the buffer
+        buffer[received++] = ch;
+
+        // Look for end-of-packet marker
+        eop = (ch == '\n');
+
+        // Check overflow condition
+        *overflow = ((received >= *size) && !eop);
+        if (*overflow) {
+            *overflow = ALLOC_CHUNK;

```

Comment:
I think a comment could be helpful here. My interpretation: we don't know how much we have left to read. We just know that the buffer is at least 1 byte too small. So we tell the caller to allocate another chunk of 128, even though that might not be enough either (or it might be more than enough).